### PR TITLE
Remove old haproxy childs after reload

### LINKED ISSUE #321: Haproxy reload not correctly in service loadbalance
we encounter the problem just like this [haproxy-issue-48](https://github.com/haproxy/haproxy/issues/48), and post a new [issue](https://github.com/haproxy/haproxy/issues/56). The old haproxy process will influence the new one that make the service visiting not stable. Until now I cannot find a solid solution and we make a little change when haproxy reload by ```-sf```. if the old haproxy process still running, then kill it by ```kill -9 <pid>```. It seems work fine but I don't know it would or not cause some bad affect when a lot of service create or modified. Did you have any suggestions or solutions that have no side effects?

### LINKED ISSUE #366: service-loadbalancer should enable the TPROXY flag
This flag is required to get the real IP address of the connection in TCP mode.
see #365

### PR DESCRIPTION
**Pull request by aledbf**
- Update haproxy to 1.6.3
- Use dumb-init to remove old haproxy child processes
  https://github.com/Yelp/dumb-init
- Fix template identation
- Remove confusing logs messages

closes #321
closes #366

### CODE REVIEW

**Comment by aledbf** [comment]
ping @bprashanth

**Comment by bprashanth** [comment]
Thanks, just 2 nits. How does this manifest at the top http-request level (other than leaked processes in the pod)?

Also we sadly can't really release a new image till https://github.com/kubernetes/kubernetes/issues/19525 is fixed (unless we revert the kubernetes godep), and I don't currently have the cycles to debug it. Maybe this week.

**Comment by aledbf** [comment]
> How does this manifest at the top http-request level (other than leaked processes in the pod)?

The "leak" is not the issue in the go code. After a reload (it could take several minutes) the old haproxy process should die (no pending connections) but it doesn't. In this scenario the old/s and the new process can reply a request and this means the configuration is old (dead containers or services missing).
I will try to create a test (a script) to reproduce this issue.

**Comment by aledbf** [comment]
```
kubectl create -f rc.yaml
kubectl create -f echoheaders.yaml
```

in a different terminal: 
`wrk -c 100 -d 120 -t 10 http://<node ip>:1936` 
(just to open connections in haproxy)

#### Scale
```
kubectl scale --replicas=2 replicationcontroller echoheaders
sleep 5
kubectl scale --replicas=10 replicationcontroller echoheaders
sleep 5
kubectl scale --replicas=1 replicationcontroller echoheaders
sleep 5
kubectl scale --replicas=10 replicationcontroller echoheaders
sleep 5
kubectl scale --replicas=5 replicationcontroller echoheaders
```

#### Results

##### servicelb:0.2 (old)
```
kubectl exec service-loadbalancer-irhp9 ps
PID   USER     TIME   COMMAND
    1 root       0:00 /service_loadbalancer --tcp-services=echoheadersx:9000
   98 root       0:02 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 91
  284 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 278
  290 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 284
  308 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 302
  338 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 332
  350 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 344
  356 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 350
  362 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 356
  380 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 374
  392 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 386
  434 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 428
  440 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 434
  464 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 458
  470 root       0:00 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 464
  480 root       0:00 ps
```
HAProxy interface reports "Statistics Report for pid 98"

##### this PR
```
$ kubectl exec service-loadbalancer-eeyty ps
PID   USER     TIME   COMMAND
    1 root       0:00 dumb-init /service_loadbalancer --tcp-services=echoheadersx:9000
    6 root       0:01 /service_loadbalancer --tcp-services=echoheadersx:9000
  382 root       1:31 haproxy -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -D -sf 376
  388 root       0:00 ps
```

HAProxy interface reports "Statistics Report for pid 382"

File: service-loadbalancer/service_loadbalancer.go [hunk 1]
```diff
@@ -619,7 +624,12 @@ func main() {
 
 	go registerHandlers(defErrorPage)
 
-	proc.StartReaper()
```
Comment (bprashanth) [inline service-loadbalancer/service_loadbalancer.go]: were we pulling in a godep for this?

File: service-loadbalancer/service_loadbalancer.go [hunk 2]
```diff
@@ -645,11 +655,11 @@ func main() {
 		glog.Fatalf("unexpected error: %v", err)
 	}
 	if !specified {
-		namespace = api.NamespaceDefault
+		namespace = api.NamespaceAll
```
Comment (bprashanth) [inline service-loadbalancer/service_loadbalancer.go]: I don't recall, so services in the default namespace will get /svcName and everything else will get /ns/svcName, correct? Can you please update the readme with this (you can do so in a follow up pr if you prefer, or I'll do it when I get a chance).

**Comment by bprashanth** [comment]
LGTM, if we can get rid of a godep that would be nice, though you can send another pr if you prefer

**Comment by bprashanth** [comment]
Let me know if you want to handle it in another pr and I'll merge this one

File: service-loadbalancer/service_loadbalancer.go (hunk 1 above)
Comment (aledbf) [inline service-loadbalancer/service_loadbalancer.go]: no, keepalived uses the same one.

**Comment by nballotta** [comment]
I tested @aledbf patch and seems working great! no more responses from different pids

**Comment by bprashanth** [comment]
@nballotta thanks! I'll be pushing out an image as soon as I've figured out https://github.com/kubernetes/kubernetes/issues/19525, which should hopefully be early next week.

**Pull request merged**